core/arch/arm/include/arm64.h

```diff
@@ -305,6 +305,21 @@
 #define ID_AA64ISAR0_SHA2_FEAT_SHA256	U(1)
 #define ID_AA64ISAR0_SHA2_FEAT_SHA512	U(2)
 
+/*
+ * Pauth keys change from atomic ones to thread ones during SP initialization.
+ * Pointer authentication must be disabled during this time or the functions
+ * will not be able to return.
```

Comment:
I'm a bit confused with this description. This means only the function tagged with attribute `__pauth` will enable BTI for PAUTH support?

Reply:
My understanding is that the macro is meant to disable PAuth only but leave BTI if it is enabled. In fact a better name would be `__nopauth`. It is a bit complex to do because the `branch-protection` attribute combines PAuth settings (`pac-ret+leaf` is what we enable in OP-TEE when `CFG_*_BTI` is enabled) as well as the BTI setting (`bti`).

The below definition isn't OK because it does not handle the core vs. TA case properly. It should not only test `CFG_CORE_BTI` and `CFG_TA_BTI` but also if the file being compiled matches those configs (`__KERNEL__`, `__LDELF__`...).

It looks like the usage is limited to kernel mode in `boot.c` and `secure_partition.c`. Therefore I suggest defining the macro in `kernel/boot.h` where it can be assumed it if for the core only:
```C
#if defined(CFG_CORE_PAUTH)
#if defined(CFG_CORE_BTI)
#define __kernel_nopauth __attribute__((target("branch-protection=bti")))
#else
#define __kernel_nopauth __attribute__((target("branch-protection=none")))
#endif
#else
#define __kernel_nopauth
#endif
```

This is still a bit ugly and I would prefer we could do differently. How is the SP case fundamentally different from the TA case?

Reply:
Yes, you are right about the intent of this macro. __nopauth is really a better name so I will rename it.
I will also remove CFG_TA_PAUTH and CFG_TA_BTI from the conditions of this macro, because you are right that they are not needed since the enable the protection for EL0.

And about your question.
In case of SP-s the problem is, that at the beginning of their initialization, the keys of the atomic context are used for PAUTH, and after the SP returns to, the kernel context keys are used. After this, when FFA messages arrive to the SP the kernel keys are used all the time both before entering EL0 and after leaving EL0. So the problem is that at init there is a switching of context (keys) which results in the PAUTH failure. 
As I see TA-s are started a bit later, when the SP-s are already initialized and kernel context is used all along, so there are no key changing.
Early TA-s are started way before enabling PAUTH or the initialization of keys.
To be honest I did not find any "rules" of the usage of atomic/kernel contexts, so I am not sure if the aforementioned behavior could be changed to avoid this issue.